### ISSUE TITLE
[Feature Request] Rename the `alias` keyword to `comptime`

### ISSUE THREAD

**Issue by augb** [modularml/mojo#171]
## Request

Rename keyword `alias` to `comptime`.

### Motivation

> **`alias`: named parameter expressions**
> It is very common to want to name compile-time values. Whereas var defines a runtime value, and let defines a runtime constant, we need a way to define a compile-time temporary value. For this, Mojo uses an alias declaration. ...

_from the [Mojo docs](https://docs.modular.com/mojo/programming-manual.html#alias-named-parameter-expressions)_

In the above definition, it is clear that an `alias` is really a `comptime` expression as opposed to strictly just another name for an expression. While it is likely there is good reason for using the keyword `alias`, I would suggest that this will lead to confusion, especially among newbies.

By using a keyword of `comptime`, the intention is clearly understood, whereas, the keyword `alias` is, while possibly technically "correct", not as clear. Aliases are typically used in other programming languages as a means of referring to a named entity (often modules) by another name. Many of these uses are in languages without the concept of compile-time expressions. In Python, for example, module aliases are created using the `as` keyword.

### Description and Requirements

The term "alias" already has a history of usage that the Mojo usage may be inadvertently going against the grain on. Using a keyword such as `comptime`, has the benefits of clearly indicating the purpose of the expression and avoiding confusion that the current keyword `alias` may engender.

**Comment by ksandvik** [comment]
What if it's just a decorator, for example the alias dtype struct could have just a decorator on top rather than multiple keywords, or same with other expressions, or an expression inside the struct. As for the name, I don't really care.

**Comment by nmsmith** [comment]
Another option would be to use a succinct sigil, e.g. `!let` and `!var`. You could use the same syntax for compile-time control flow, e.g. `!if`. (The `@parameter` decorator is a bit too noisy imo.)

**Comment by lattner** [comment]
comptime is really obvious to Zig folk, but that's not really our audience.  You're right that alias may not be the right word to use here either.  Aligning this around "parameter" could be a good way to go, but I'm curious if there are other suggestions.

Once nice thing about "alias" is that it is more obvious for the trivial cases like `alias my_magic = 12312` or `alias Int8 = SIMD[DType.si8, 1]`.    That doesn't make it the right thing, but it is a nice thing.

**Comment by augb** [comment]
> comptime is really obvious to Zig folk, but that's not really our audience. You're right that alias may not be the right word to use here either. Aligning this around "parameter" could be a good way to go, but I'm curious if there are other suggestions.

I think my point is not to copy Zig, rather, it is to make clear that the value is compile-time value. I'm open to other suggestions, but I think it might be better err on the side of clarity.

> Once nice thing about "alias" is that it is more obvious for the trivial cases like `alias my_magic = 12312` or `alias Int8 = SIMD[DType.si8, 1]`. That doesn't make it the right thing, but it is a nice thing.

`alias` suggests that it could be at runtime (or compile-time, as it is currently used). While I agree that your examples make perfect sense, these are still compile-time expressions, if I am understanding the docs correctly. The keyword `alias` gives no indication of this, neither does aligning around "parameter".

I could see the possibility of a future need for the keyword `alias` that can be used in a runtime-only context. Current use excludes that possibility.

> Aligning this around "parameter" could be a good way to go, but I'm curious if there are other suggestions.

@lattner, could you elaborate on what you mean by this?

**Comment by lattner** [comment]
+1, I wasn't arguing for alias, just explaining the though process.

> Aligning this around "parameter" could be a good way to go, but I'm curious if there are other suggestions.

Again, I see problems with it, but if we replaced the keyword "alias x = 42" with "parameter x = 42", then we can say "it's a declaration of a parameter" and that "parameters are all compile time expressions.

**Comment by nmsmith** [comment]
No love for the `!let` (or any other sigil) idea I mused about above? 🥲

**Comment by lattner** [comment]
> No love for the !let (or any other sigil) idea I mused about above? 🥲

Not really, alias (regardless of what it is called) is a declaration of a thing.  We need spoken vocabulary for programmers to describe these things.  It isn't just about encoding things in source code for the compiler, it is allowing humans to communicate ideas as well.

Also, "let" values are not aliases.  They've very different.  A let isn't mutable after it is initialized, which is a flow sensitive property, e.g. this is allowed:

```
  let x : Int 
  if cond:
    x = foo()
  else:
    x = bar()
  use(x)
```

which isn't allowed for aliases.

**Comment by nmsmith** [comment]
I actually 100% agree with you that verbalizability is important. The project I'm working on right now is developing a PL with a natural language-ish syntax. I'm a huge fan of English keywords as opposed to sigils. I need no convincing!

Despite my love of English keywords, I suggested using a sigil in an attempt to address a challenge that I see with Mojo's metaprogramming syntax. The challenge being:
  - When multiple program constructs (e.g. `alias` and `if`, and I'm assuming `for` at some point) can run at compile-time, you don't want to have to give each of those control constructs a unique name (e.g. `compileTimeIf` and `compileTimeFor`). That's what `@parameter` is designed to prevent. It just modifies the interpretation of an existing keyword. So it would be nice to have a syntax that can be used across all of Mojo's constructs to indicate that you are invoking the compile-time version of it.

Regarding that code sample that you say is not supported for aliases: have you considered supporting it? Having the same semantics at compile-time and run-time would reduce the number of concepts that people need to learn to do metaprogramming. (The `if` in that example would have to be a compile-time `@parameter if`, of course.)

If you had a sigil such as `!` or `~` or whatever, you'd just write the exact same thing, but use the sigil to request compile-time execution:
```
  ~let x : Int 
  ~if cond:
    x = foo()
  ~else:
    x = bar()
```

I'm only proposing a sigil because keywords don't seem to work very well:

```
  comp let x : Int 
  comp if cond:
    x = foo()
  else:
    x = bar()
```

The end result is having fewer concepts to learn. There's no `alias`, there's just `let`.

**Comment by dom96** [comment]
I don't have any strong feelings, but just wanted to mention how another language does it. Nim uses `const` for this. Developers quickly learn that `const` means "compile-time", the same will probably happen with `alias` but if you want to reuse something another language already uses then `const` might be a good way to go.

That said, re-reading the docs for `alias` I get the feeling that it actually is strictly an alias and thus the `alias` keyword makes sense for it. Does the compiler forcefully evaluate expressions in an alias at compile-time? i.e. will `alias x = 5 + aVariable; print(x)` fail at compile-time because `aVariable` is not known at compile-time or is this simple expression replacement and `x` is just filled with `5+aVariable` in the `print`? In Nim `const x = 5 + aVariable; echo(x)` doesn't compile.

**Comment by augb** [comment]
> I'm just saying that if we replaced the keyword "alias x = 42" with "parameter x = 42", then we can say "it's a declaration of a parameter" and that "parameters are all compile time expressions."

@lattern, `parameter` gets closer, but I still think it is too overloaded in programming circles. After all, parameters are in the call signatures of a vast number of function definitions.

> **alias: named parameter expressions**

From the docs, "named parameter expressions" are really "named (compiler) parameter expressions", aren't they? By using `parameter` for a narrow meaning such as this, I think we are adding to the cognitive burden of a developer.

> ... Nim uses const for this. Developers quickly learn that const means "compile-time", the same will probably happen with alias but if you want to reuse something another language already uses then const might be a good way to go.

@dom96 I like the idea of using `const`. It aligns with the concept of an unchanging value. C#, Rust, and many others already have the `const` keyword. Having this determined at compile-time doesn't throw any significant hurdles in the way, as far as I am concerned.

If `const` is problematic for reasons that are not immediately apparent to me, I would argue for `compexp` (instead of `comptime`). `compexp` = "compiler expression". I think this pretty clearly identifies what follows.

> Does the compiler forcefully evaluate expressions in an alias at compile-time?

@dom96, my understanding is `alias` is _always_ about compile-time named parameter expressions.

**Comment by Janrupf** [comment]
It might even make sense to be able to just mark expressions and bindings as compiletime. In this case a keyword may be the better choice. With this it may look like the following:

```
comptime SOMETHING = 1234

if comptime(some_fn()):
    do_abc()
else:
    do_xyz()
    
some_call(SOMETHING, comptime(do_it()), get_user_input())
```

Reasoning here would be, that it is more flexible and allows partially marking things as compiletime. This however assumes that a reasonable optimizer is in place, which can detect a compiletime static condition and perform dead code elimination accordingly.

In this case it becomes questionable whether it makes sense to mark functions as compiletime (or even provide that ability). The call-site would control whether the called code is evaluated at compiletime, and if so, everything down the call tree is. At definition-site (as in, the fn declaration) it may then only make sense to mark it as compiletime if it should be required to be called in a compiletime context.

Regarding the naming, I think something like `const` or `comptime` would be a good idea, since it has a very clear intention. I like `comptime` the most out of all the proposed names so far, since it very clearly states what it does.

**Comment by nmsmith** [comment]
I updated my post above with a more detailed proposal for Mojo's metaprogramming syntax. In short: I think it's worth focusing on finding a clear, and easy-to-understand _model_ for compile-time execution. Once this is done, it will be easier to figure out a good name for specific constructs (such as `alias`).

By focusing too narrowly on the name for the `alias` construct, we may be missing the forest for the trees.

**Comment by augb** [comment]
To summarize my view at this point in the discussion, I offer the following:

I started this as a replacement for the keyword `alias`; however, I'm seeing the benefit of aligning `@parameter` and `alias` under a single keyword such as `const`, `compexp` or `comptime`.

While I understand the choice of `alias` and `@parameter`, respectively, I feel these are too overloaded with meaning from other languages and common usage of these terms in programming.

If possible, having a single keyword, as opposed to a mix of a keyword and a decorator, would be preferable.

**Comment by Mogball** [comment]
This is veering off into discussion land. An issue is not the right place for something like this. Please start a new discussion (via Github discussions in this same repo) if you would like to continue this conversation.

Also FYI we already have compile-time ifs, in the form of `@parameter if a + b`

**Comment by nmsmith** [comment]
If Github issues aren't the appropriate place to discuss language features, then perhaps the "Feature Request" option should be removed from the list of issue templates. Because otherwise, I predict discussions such as this will be commonplace. (Because people are obviously going to respond to Feature Request issues with their thoughts about the feature.)

**Comment by Mogball** [comment]
That's probably a good idea. That said, certain feature requests are less controversial than others. I suspect this one should have been "[RFC] Rename the `alias` keyword"

**Comment by nmsmith** [comment]
Perhaps the template should be "Missing feature" instead then. To stop people requesting a redesign of things that already exist.

**Comment by Mogball** [comment]
Makes sense me to @goldiegadde. Although the problem will not go away. I suspect there will be more "please move this to an RFC" in the future.

**Comment by augb** [comment]
Discussion started in #190.
